examples/js/loaders/GLTFLoader.js

```diff
@@ -1860,43 +2226,62 @@ THREE.GLTFLoader = ( function () {
 
 		var parser = this;
 
-		return this.getDependency( 'texture', mapDef.index ).then( function ( texture ) {
+		return this._onBefore( 'map', mapDef ).then( function ( def ) {
```

Comment:
Do we need this hook if we move texture format out to `assignTexture`? It seems like we could move the `KHR_texture_transform` functionality out to a `onAfterTexture` hook. I think the existing texture hooks should handle all of the cases you are using the map hook for.

Reply:
I think we still need. `assignTexture` handles `mapDef` ([`textureInfo` in glTF](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#textureinfo)) which has a reference to texture and can have `KHR_texture_transform` extension, while `load/on/onBefore/onAfterTexture` handles `textureDef` ([`texture` in glTF](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#texture)) which defines texture itself.

texture itself can't know whether `KHR_texture_transform` is applied to it.

Reply:
Because `KHR_texture_transform` is nested underneath a material object, I think I would expect it to use onAfterMaterial to apply transforms to textures. We might be able to do without the `textureInfo` or `map` callbacks here?

Reply:
Let me confirm one thing.

In the current proposed API, extension handler plugin should have name and its `on/onBefore/onAfterXXX` (XXX is Material, Texture, Mesh, or so on) is fired if a XXX asset definition has the extension whose name is same as the extension handler plugin's.

`KHR_texture_transform` extension is placed in `textureInfo`, so `on/onBefore/onAfterTextureInfo` can be fired but `on/onBefore/onAfterMaterial` won't be. Do you want to change it, `on/onBefore/onAfterMaterial` can be fired if underneath `textureInfo` has a certain extension?

Reply:
Oh I see... I think plugins should override whatever on/onBefore/onAfter hooks they need to use. If a plugin comes along that needs to use both `onBeforeCamera` and `onAfterMaterial` for whatever reason, that's fine.

In other words, the plugin should decide when it gets called, not the location of the extension data in the glTF schema. Two ways to implement this would be:

1. Loop over all registered plugins to see if they have onBeforeFoo method, calling those that do.
2. Each plugin has a `.hooks = ['texture', ...]` array defining the resources it wants to affect. So unlit might have this:

```js
class UnlitExtension {
  constructor {
    this.name = 'KHR_materials_unlit';
    this.hooks = [ 'material' ];
  }
  onAfterMaterial () { ... }
}
```

The KHR_texture_transform material could then use the `material` hook, or a new `textureInfo` hook as you've done here, and either would would work. But ideally, extensions in new places (like `material.pbrMetallicRoughness.*`) would not always require us to add a new set of three callbacks.

Reply:
Hm, the reason why I added that limitation is likely we can implement extension handlers of the existing extensions we currently support even with the limitation and I thought it can make the API and implementation simpler, but yeah it may be important for us to have more flexible API.

> 1. Loop over all registered plugins to see if they have onBeforeFoo method, calling those that do.
> 2. Each plugin has a .hooks = ['texture', ...] array defining the resources it wants to affect. 

I guess 1 can have performance impact if a lot of plugins are registered so am inclined for 2.

Or if we think defining `on(Before|After)XXX` method and `hooks = [ XXX ]` property sounds duplicated, we may automatically generate `hooks` (or similar stuffs), for example when a plugin is registered, by checking what method is defined. I guess checking up to tens of methods one time per a plugin may not hurt.

Reply:
Let me confirm my understanding. With this example

```javascript
class UnlitExtension {
  constructor {
    this.name = 'KHR_materials_unlit';
    this.hooks = [ 'material' ];
  }
  onAfterMaterial () { ... }
}
```

`onAfterMaterial` is called regardless of whether material definition has `KHR_materials_unlit` extension so `onAfterMaterial` checks if the definition has the extension and does nothing if it doesn't have the extension, correct?

```javascript
  onAfterMaterial(material, materialIndex, materialDef, parser) {
    if (materialDef.extensions.KHR_material_unlit === undefined) {
      return material;
    }
    ...
  }
```

Reply:
BTW

> Because KHR_texture_transform is nested underneath a material object, I think I would expect it to use onAfterMaterial to apply transforms to textures.

If I understand correctly, isn't it difficult to handle `KHR_texture_transform` in `onAfterMaterial`? It may be impossible for `onAfterMaterial` to find a mapping from a certain `textureInfo` to a certain Three.js material property because the mapping is core/extension or plugin specific.

For example imagine that there is the following material definition and `Foo_material_extension` handler plugin assigns `fooTexture` to Three.js material `barMap`. I don't think `KHR_texture_transform` plugin's `onAfterMaterial` can know the mapping from `fooTexture` `textureInfo` to Three.js material `barMap`.

```
"material": [{
  "extensions": {
    "Foo_material_extension": {
      "fooTexture": { // -> Three.js material.barMap
        "index": 0,
        "extensions": {
          "KHR_texture_transform": {
            ...
          }
        }
      }
    }
  }
}]
```

Reply:
Updated. I have removed the limitation and `on/onBefore/onAfterXXX` is always fired if theyre refined regardless of extension definition. So extension handler plugin needs to check extension definition in plugin side to know whether the plugin should process.

>> 1. Loop over all registered plugins to see if they have onBeforeFoo method, calling those that do.
>> 2. Each plugin has a .hooks = ['texture', ...] array defining the resources it wants to affect.
>>
> I guess 1 can have performance impact if a lot of plugins are registered so am inclined for 2.

I said I'm inclined for 2. but I implemented 1. so far because 1. is easier to implement and I don't see significant performance (response time) degradation for now. Let's revisit if we realize it can have a big performance impact (in some cases).

And regarding where we should handle `KHR_texture_transform`, I'd like to discuss in more detail in another PR moving `KHR_texture_transform` handler into the new API, and I'd like to keep `textureInfo` hook point as is for now.